### PR TITLE
Allow public invites for alice/faber demo

### PR DESCRIPTION
**Pull request by ianco**
Signed-off-by: Ian Costanzo <ian@anon-solutions.ca>

Addresses https://github.com/hyperledger/aries-cloudagent-python/issues/1571

Note - added a `--reuse-connections` parameter to faber - this will trigger faber to use a public did in the didexchange connection (and alice will always "reuse connection" if possible) and the connection indeed gets reused, however right now there are no webhooks for the "reuse" messages so the alice/faber controllers are getting hung ...

### CODE REVIEW

**Comment by swcurran** [comment]
Hmm...I ran this:

```
./run_demo faber --public-invites
Preparing agent image...
sha256:94c44484e255f32387e52e2b69f8f7e767625c3a36235cd1239b5481dbb67cc9
Trying to detect ngrok service endpoint
ngrok not detected for agent endpoint
172.17.0.1
Starting [faber] agent with args [--port 8020 --public-invites]
usage: faber.py [-h] [--no-auto] [-p <port>] [--did-exchange] [--revocation]
                [--tails-server-base-url <tails-server-base-url>]
                [--cred-type <cred-type>] [--aip <api>] [--timing]
                [--multitenant] [--mediation] [--multi-ledger]
                [--wallet-type <wallet-type>]
                [--endorser-role <endorser-role>] [--arg-file <arg-file>]
faber.py: error: unrecognized arguments: --public-invites
```

**Comment by ianco** [comment]
> Hmm...I ran this:
> 
> ```
> ./run_demo faber --public-invites
...
> faber.py: error: unrecognized arguments: --public-invites
> ```

You don't need to add a new parameter to faber, the `--public-invites` param is added to the aca-py agent that faber (and alice) start up under the hood

**Comment by swcurran** [comment]
So it works, but each time I do a new invitation, I get a new connection on both sides.  That means that reuse is not being used, which is the goal.  Alice should be responding with a "reuse" message, and her "new" connection should be removed.  Likewise on Faber -- once she responds with the reuse, the hanging connection should be removed.  This (pretty likely :-) ) works because there was a follow up tweak to DID Exchange to do the clean up after someone tried this out.

Thoughts?

**Comment by swcurran** [comment]
I think the reuse _should_ be automatic, handled by ACA-Py, but am not certain how it works.

**Comment by swcurran** [comment]
Also, I noticed that the "their_public_did" item on the connection object was not set, which is a bit odd.  Not sure why it wouldn't be, but I think that is a pre-requisite for "reuse" on a connection.

**Comment by ianco** [comment]
I believe the auto-reuse only works when you use public DIDs for connections, which the demo is not doing

**Comment by ianco** [comment]
> I believe the auto-reuse only works when you use public DIDs for connections, which the demo is not doing

I guess it depends on the intent, I though we were just trying to unblock Peter (issue https://github.com/hyperledger/aries-cloudagent-python/issues/1571 where he is using the API).

If we want to configure alice/faber to use connection reuse for the demo then there is more work required.  Let me know ...

**Comment by swcurran** [comment]
Let's leave it for now.  Peter's goal is to enable connection reuse, so that's really what we are after. 

I think it would be nice to have the demo optional support connection reuse, so it would have an optional parameter, making it easy to see what the controller has to do to support reuse.  How about we leave this as a draft PR and when you get a chance, you can update it?  I'm hoping it is a pretty small change.

**Comment by ianco** [comment]
> Let's leave it for now. Peter's goal is to enable connection reuse, so that's really what we are after.
> 
> I think it would be nice to have the demo optional support connection reuse, so it would have an optional parameter, making it easy to see what the controller has to do to support reuse. How about we leave this as a draft PR and when you get a chance, you can update it? I'm hoping it is a pretty small change.

Sure that works, I've converted the PR to draft.

**Comment by ianco** [comment]
Fixed demo - run with `./run_demo faber --reuse-connections` - this will use faber's public DID for the connection and allow connection reuse.

(Alice will always try to reuse connections if possible.)

File: demo/runners/agent_container.py [hunk 1]
```diff
@@ -96,8 +96,23 @@ def connection_ready(self):
         return self._connection_ready.done() and self._connection_ready.result()
 
     async def handle_oob_invitation(self, message):
+        print("handle_oob_invitation()")
         pass
 
+    async def handle_connection_reuse(self, message):
```
Comment (shaangill025) [inline demo/runners/agent_container.py:102]: Is this function being used?

File: demo/runners/agent_container.py [hunk 2]
```diff
@@ -96,8 +96,23 @@ def connection_ready(self):
...
+        # we are reusing an existing connection, set our status to the existing connection
+        if not self._connection_ready.done():
+            self.connection_id = message["connection_id"]
+            self.log("Connected")
+            self._connection_ready.set_result(True)
+
+    async def handle_connection_reuse_accepted(self, message):
```
Comment (shaangill025) [inline demo/runners/agent_container.py:109]: Same question as above.

File: demo/runners/agent_container.py (hunk 1 above)
Comment (ianco) [reply demo/runners/agent_container.py:102]: This allows the controller (faber) to tag the active connection (handles the webhook)

File: demo/runners/agent_container.py (hunk 2 above)
Comment (ianco) [reply demo/runners/agent_container.py:109]: This allows the controller (alice) to tag the active connection (handles the webhook)

**Review by shaangill025 — APPROVED** [review]
`Alice` reuses the existing connection as intended.

**Pull request merged**